eth/vm/gas_meter.py

```diff
@@ -16,17 +17,49 @@
 )
 
 
+def default_refund_strategy(meter: "GasMeter", amount: int) -> None:
+    if amount < 0:
+        raise ValidationError("Gas refund amount must be positive")
+
+    meter.gas_refunded += amount
+
+    meter.logger.trace(
+        'GAS REFUND: %s + %s -> %s',
+        meter.gas_refunded - amount,
+        amount,
+        meter.gas_refunded,
+    )
+
+
+def allow_negative_refund_strategy(meter: "GasMeter", amount: int) -> None:
+    meter.gas_refunded += amount
```

Comment:
So it raises a flag for me when there is a mutual dependency (`GasMeter` on the refund strategy and refund strategy on GasMeter).

One alternative:
```py
def allow_negative_refund_strategy(previous_refund_total: int, amount: int) -> None:
    return previous_refund_total + amount
```

Which makes the strategy both stateless and side-effect-free.

Reply:
I also thought about that but it is limiting the usefulness heavily. E.g. what if you want to implement a strategy that throws if e.g. `gas_refunded` or `gas_remaining` is above some treshold (just making things up). Sure, we could pass these as well but at this point we may just pass the entire meter and let the function inspect whatever it wants?

Reply:
Ok, I'm not married to that particular alternative, but the mutual dependency seems fairly undesirable. If there are no other options come to mind, then we can just drop the compositional approach.

Reply:
Well ok, maybe designing this in the most flexible way shouldn't be our main priority for now. We can design it in the stateless way you suggest as it is flexible enough to fulfill our current needs and if we ever come across the need for an alternative refund strategy that this system can't handle we can reconsider.